### ISSUE TITLE
DigitalOcean example?

### ISSUE THREAD

**Issue by jbonnett92** [ctrox/csi-s3#16]
Hi,
Does anyone have a DigitalOcean example?
I can't seem to get it to work with it.

Thanks,
Jamie

**Comment by ctrox** [comment]
I do not have any experience with DigitalOcean myself but it should just depend on what [Kubernetes/CSI version](https://kubernetes-csi.github.io/docs/introduction.html#kubernetes-releases) is supported.
If they already support Kubernetes 1.13 I can recommend you try the new release (v1.1.0) that I just uploaded, which should make things a bit easier to deploy.

**Comment by jbonnett92** [comment]
I'm not running the CSI on DigitalOcean but rather trying to connect the CSI to DigitalOceans S3 storage service.

**Comment by ctrox** [comment]
Ah right, their storage should be interoperable with the S3 API so it should "just work". Have you tried just setting access/secret keys and the endpoint to e.g. `nyc3.digitaloceanspaces.com` while setting `region: ""`?

**Comment by jbonnett92** [comment]
I think it is just me, I did try setting it up with AWS in the end and I am also having issues there:
```I0516 02:08:11.573841       1 event.go:221] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"csi-s3-pvc", UID:"3adf8529-777d-11e9-9cb4-00163ca95b8f", APIVersion:"v1", ResourceVersion:"8313468", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "csi-s3": rpc error: code = Unknown desc = failed to create volume pvc-3adf8529-777d-11e9-9cb4-00163ca95b8f: The provided 'x-amz-content-sha256' header does not match what was computed.```

**Comment by pjanuario** [comment]
@jbonnett92 I recently started using and testing it with DO and followed this post https://medium.com/asl19-developers/create-readwritemany-persistentvolumeclaims-on-your-kubernetes-cluster-3a8db51f98e3 

Using the following credentials:

```
---
apiVersion: v1
kind: Secret
metadata:
  name: csi-s3-secret
stringData:
  accessKeyID: "XXXXX"
  secretAccessKey: "XXXX"
  endpoint: "https://ams3.digitaloceanspaces.com"
  region: ""
```

I had a similar problem to yours that passed by removing this from the secret:

`encryptionKey: ""`

Also I didnt needed to perform step 4, things were just working without issues.

Just a short notice beware that storage class reclaim policy is to default wich is "delete" and while using helm to deploy a chart that contains the pvc it ended in deleting the pvc and pv, this means the DO object space is deleted.

**Comment by jbonnett92** [comment]
@pjanuario I found this article too, it didn't work for me.

**Comment by pjanuario** [comment]
@jbonnett92 if you followed the article becarefull, that their example contains the encryption key, your symptoms are similar to mine and removing that key from the secret resolved the problem.

**Comment by jbonnett92** [comment]
@pjanuario I did do that and it still didn't work.

**Comment by pjanuario** [comment]
@jbonnett92 not sure if it helps but I am using the yaml from this repo and  the tag `v1.1.0` and I run k8 1.13 (should not have impact I think)

**Comment by Matzu89** [comment]
I'm using this with DO Spaces and the claim gets through blazingly fast, so that works. 
Only problem is I get a `Permission denied` from every container that is using it.

Anyone experienced the same problem? How did you solve it?

**Comment by jbonnett92** [comment]
@Matzu89 I never solved my problem.

**Comment by pjanuario** [comment]
@jbonnett92 and @Matzu89 the setup is working for me without any issues so fare. I have several spaces mounted as volumes and use it with different applications, mostly to store uploaded images.

**Comment by Berndinox** [comment]
Tried the example of @pjanuario today: DO K8s and DO spaces. Worked like a charm!

**Comment by jbonnett92** [comment]
@Berndinox @pjanuario I am running bare metal, I find this effects the setup of everything for some reason. Would it affect this CSI?

**Comment by pjanuario** [comment]
@jbonnett92 I can't answer to that with certainty, but I would say that it should not since this is mapping volumes and data fetch is done with API calls, but maybe there are some details there...
Sorry I can't help that much.

**Comment by Berndinox** [comment]
@jbonnett92 may fire up an DO managed K8s and try it on there. Maybe you are missing so pre-requisits DO is already configuring for you?!

**Comment by jbonnett92** [comment]
Ok I tried setting it up on Do and the PVC just gets stuck in pending, here is my command line output:
```
NAME           READY   STATUS    RESTARTS   AGE
csi-s3-6rxc7   2/2     Running   0          22m
Jamies-MBP:kubernetes Jbonnett$ kubectl -n kube-system logs csi-s3-6rxc7 -c csi-s3
I0723 00:15:03.161255       1 s3-driver.go:80] Driver: ch.ctrox.csi.s3-driver 
I0723 00:15:03.161360       1 s3-driver.go:81] Version: v1.1.1 
I0723 00:15:03.161371       1 driver.go:81] Enabling controller service capability: CREATE_DELETE_VOLUME
I0723 00:15:03.161378       1 driver.go:93] Enabling volume access mode: SINGLE_NODE_WRITER
I0723 00:15:03.162145       1 server.go:108] Listening for connections on address: &net.UnixAddr{Name:"//csi/csi.sock", Net:"unix"}
I0723 00:15:03.317947       1 utils.go:97] GRPC call: /csi.v1.Identity/GetPluginInfo
I0723 00:15:04.177154       1 utils.go:97] GRPC call: /csi.v1.Node/NodeGetInfo
Jamies-MBP:kubernetes Jbonnett$ kubectl get pvc csi-s3-pvc
NAME         STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
csi-s3-pvc   Pending                                      csi-s3         14m
Jamies-MBP:kubernetes Jbonnett$ kubectl logs -l app=csi-provisioner-s3 -c csi-s3
Jamies-MBP:kubernetes Jbonnett$ kubectl logs -l app=csi-s3 -c csi-s3
Jamies-MBP:kubernetes Jbonnett$ 

```

**Comment by Berndinox** [comment]
Seems like the DO Space can not be created, can you verify this?
If there is no space called pvc-blahblah123, i would look onto the endpoint and api Key Config

**Comment by jbonnett92** [comment]
Yes you're right, the config is literally a copy and paste job I can't see that being wrong. Especially when I have tried implementing this more than twice.

**Comment by jbonnett92** [comment]
On the non DO one I get this:
```
I0723 22:55:03.407302       1 controllerserver.go:59] Got a request to create volume pvc-1610f063-ad96-11e9-9cb4-00163ca95b8f
E0723 22:55:03.536025       1 utils.go:101] GRPC error: failed to check if bucket pvc-1610f063-ad96-11e9-9cb4-00163ca95b8f exists: The specified key does not exist.
```

**Comment by jbonnett92** [comment]
Ok I finally have it working, although `ReadWriteMany` is any ideas? I am using the goofys mounter stated by @pjanuario and the article posted says that `ReadWriteMany` works, although if multiple pods try accessing the PVC it just fails to attach.

**Comment by jbonnett92** [comment]
Weirdly enough deleting and applying the configs again worked

**Comment by jbonnett92** [comment]
Sorry to be a pain, I keep getting this error with different pods:
```
pod has unbound immediate PersistentVolumeClaims (repeated 3 times)
Back-off restarting failed container
```
this happens with many different deployments.
I did test with the pod example in the readme, it worked and then when I try adding a new deployment it fails.

**Comment by polRk** [comment]
Why i can't see files in my DO spaces ?

**Comment by jbonnett92** [comment]
> Why i can't see files in my DO spaces ?

As far as I remember, the issue that I had was that I was including the region in the endpoint. I will double check the config later and I will let you know!

**Comment by polRk** [comment]
This is my secret file (deployed to kube-system)
```
apiVersion: v1
kind: Secret
metadata:
  name: csi-s3-secret
stringData:
  accessKeyID: "MySuperSecretKey"
  secretAccessKey: "MySuperSecretAccessKey"
  endpoint: https://fra1.digitaloceanspaces.com
  region: ""
  encryptionKey: ""
```
This is my storageclass.yaml
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: csi-s3
provisioner: ch.ctrox.csi.s3-driver
parameters:
  # specify which mounter to use
  # can be set to rclone, s3fs, goofys or s3backer
  mounter: goofys
  csi.storage.k8s.io/provisioner-secret-name: csi-s3-secret
  csi.storage.k8s.io/provisioner-secret-namespace: kube-system
  csi.storage.k8s.io/controller-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/controller-publish-secret-namespace: kube-system
  csi.storage.k8s.io/node-stage-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-stage-secret-namespace: kube-system
  csi.storage.k8s.io/node-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-publish-secret-namespace: kube-system
```
```

**Comment by jbonnett92** [comment]
Try removing the encryption key's key.
Also make sure you are not using the personal access token and key for this.
Also make sure that the secret is in the kube-system namespace.

Apart from that I can't tell of anything wrong.

**Comment by polRk** [comment]
> Try removing the encryption key's key.

Yes, I tried, the result is the same

> Apart from that I can't tell of anything wrong.

:(

**Comment by jbonnett92** [comment]
What about the other two?

**Comment by polRk** [comment]
> Also make sure that the secret is in the kube-system namespace
```
csi-attacher-sa-token-kwmk4        kubernetes.io/service-account-token   3      4h38m
csi-do-controller-sa-token-j6tqb   kubernetes.io/service-account-token   3      44d
csi-do-node-sa-token-mv89r         kubernetes.io/service-account-token   3      44d
csi-provisioner-sa-token-sdkgt     kubernetes.io/service-account-token   3      4h38m
csi-s3-secret                      Opaque                                5      4h39m
csi-s3-token-rnnk4                 kubernetes.io/service-account-token   3      4h37m
```

**Comment by jbonnett92** [comment]
That isn't showing me a namespace?

**Comment by polRk** [comment]
> What about the other two?

Yes, i using personal access token (

**Comment by polRk** [comment]
> That isn't showing me a namespace?

`➜  ~ kubectl get secret -n kube-system`

**Comment by jbonnett92** [comment]
> > What about the other two?
> 
> Yes, i using personal access token (

There is your problem, you should be using a spaces token.

**Comment by polRk** [comment]
> There is your problem, you should be using a spaces access keys.

After updating i got 
```
name-pvc                Pending                                                                        csi-s3             2m26s
```

**Comment by jbonnett92** [comment]
Anything in DO?

**Comment by polRk** [comment]
```
➜  kubernetes kubectl get secret csi-s3-secret  -o yaml
apiVersion: v1
data:
  accessKeyID: KEY
  encryptionKey: ""
  endpoint: aHR0cHM6Ly9mcmExLmRpZ2l0YWxvY2VhbnNwYWNlcy5jb20=
  region: ""
  secretAccessKey: KEY
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Secret","metadata":{"annotations":{},"name":"csi-s3-secret","namespace":"default"},"stringData":{"accessKeyID":"KEY","endpoint":"https://fra1.digitaloceanspaces.com","region":"","secretAccessKey":"KEY"}}
  creationTimestamp: "2019-11-04T09:46:33Z"
  name: csi-s3-secret
  namespace: default
  resourceVersion: "5752632"
  selfLink: /api/v1/namespaces/default/secrets/csi-s3-secret
  uid: 9bae5333-f5af-40c8-96aa-3f9b43102f65
type: Opaque
```

**Comment by jbonnett92** [comment]
Your secret isn’t in kube-system, it’s in default.

**Comment by polRk** [comment]
> Your secret isn’t in kube-system, it’s in default.

I have secrets in kube-system and in default

**Comment by jbonnett92** [comment]
It won’t see the one in default.

**Comment by polRk** [comment]
```
➜  kubernetes kubectl get secret csi-s3-secret  -o yaml -n kube-system
apiVersion: v1
data:
  accessKeyID: KEY
  encryptionKey: ""
  endpoint: aHR0cHM6Ly9mcmExLmRpZ2l0YWxvY2VhbnNwYWNlcy5jb20=
  region: ""
  secretAccessKey: KEY
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Secret","metadata":{"annotations":{},"name":"csi-s3-secret","namespace":"kube-system"},"stringData":{"accessKeyID":"KEY","endpoint":"https://fra1.digitaloceanspaces.com","region":"","secretAccessKey":"KEY"}}
  creationTimestamp: "2019-11-04T09:51:27Z"
  name: csi-s3-secret
  namespace: kube-system
  resourceVersion: "5754056"
  selfLink: /api/v1/namespaces/kube-system/secrets/csi-s3-secret
  uid: 43a94988-0fe7-4357-b807-9d0adbb527e5
type: Opaque
➜  kubernetes kubectl get secret csi-s3-secret  -o yaml
apiVersion: v1
data:
  accessKeyID: KEY
  encryptionKey: ""
  endpoint: aHR0cHM6Ly9mcmExLmRpZ2l0YWxvY2VhbnNwYWNlcy5jb20=
  region: ""
  secretAccessKey: KEY
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Secret","metadata":{"annotations":{},"name":"csi-s3-secret","namespace":"default"},"stringData":{"accessKeyID":"KEY","endpoint":"https://fra1.digitaloceanspaces.com","region":"","secretAccessKey":"KEY"}}
  creationTimestamp: "2019-11-04T09:46:33Z"
  name: csi-s3-secret
  namespace: default
  resourceVersion: "5754061"
  selfLink: /api/v1/namespaces/default/secrets/csi-s3-secret
  uid: 9bae5333-f5af-40c8-96aa-3f9b43102f65
type: Opaque
```

**Comment by jbonnett92** [comment]
Remove the one in default and add the spaces access key to the kube-default one

**Comment by polRk** [comment]
I delete and install storageclass
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: csi-s3
provisioner: ch.ctrox.csi.s3-driver
parameters:
  # specify which mounter to use
  # can be set to rclone, s3fs, goofys or s3backer
  mounter: goofys
  csi.storage.k8s.io/provisioner-secret-name: csi-s3-secret
  csi.storage.k8s.io/provisioner-secret-namespace: kube-system
  csi.storage.k8s.io/controller-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/controller-publish-secret-namespace: kube-system
  csi.storage.k8s.io/node-stage-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-stage-secret-namespace: kube-system
  csi.storage.k8s.io/node-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-publish-secret-namespace: kube-system
```
Pod is successfully running. But i don't see any files
![cLWU8uuYZu0](https://user-images.githubusercontent.com/39828645/68130619-b6add400-ff2c-11e9-810e-89e7e3fd3a41.jpg)

**Comment by jbonnett92** [comment]
Ohh it is creating the PVC in DO, ok are you sure that what is using the PVC is creating files?

**Comment by jbonnett92** [comment]
I see two items which is metadata and a folder, which means it should be working.

**Comment by polRk** [comment]
> Ohh it is creating the PVC in DO, ok are you sure that what is using the PVC is creating files?

I don't know. I'm not sure.

pvc.yaml
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: {{ include "chart.fullname" . }}
  labels:
{{ include "chart.labels" . | indent 4 }}
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: {{ .Values.pvc.storage }}
  storageClassName: csi-s3

```
deployment.yaml
```
          volumeMounts:
            - name: {{ include "chart.fullname" . }}-volume
              mountPath: "/var/wwww/html/wp-content/uploads"
      volumes:
        - name: {{ include "chart.fullname" . }}-volume
          persistentVolumeClaim:
            claimName: {{ include "chart.fullname" . }}
```

**Comment by jbonnett92** [comment]
Ohh hold on, are you using the CSI for the database by chance? You need a fast form of storage to do database operations. If not then ok, also I don't think Wordpress creates any files until you upload them to the site e.g. create a post with an image.

**Comment by polRk** [comment]
> Ohh hold on, are you using the CSI for the database by chance? You need a fast form of storage to do database operations. If not then ok, also I don't think Wordpress creates any files until you upload them to the site e.g. create a post with an image.


Yes, I want to store wordpress photos not in digitalocean volumes, but in spaces

**Comment by jbonnett92** [comment]
How did you install Wordpress? Helm? Do you have the command you used?

**Comment by polRk** [comment]
> How did you install Wordpress? Helm? Do you have the command you used?

Build docker image
Base on https://github.com/polRk/docker-wordpress
```
FROM polrk/wordpress
COPY wordpress/ /var/www/html/
```

And apply 
helm upgrade domain-$BRANCH $GITHUB_WORKSPACE/.helm --install --set image.tag=$TAG,ingress.host=$BRANCH.domain.com

**Comment by jbonnett92** [comment]
I would recommend using the helm chart https://github.com/helm/charts/tree/master/stable/wordpress

At least for now, you can set the persistent storage class (e.g. images) and the one for the database.

**Comment by polRk** [comment]
> I would recommend using the helm chart https://github.com/helm/charts/tree/master/stable/wordpress
> 
> At least for now, you can set the persistent storage class (e.g. images) and the one for the database.

I using external mariadb database. I need spaces only for wordpress uploads folder.

https://github.com/helm/charts/tree/master/stable/wordpress
It's doesn't working with feature-development. And I need the uploads folder to be the volume

**Comment by polRk** [comment]
> It's doesn't working with feature-development. 

helm list 
```
domain-branch1
domain-branch2
...

domain-branch99
```
I need immutable docker image this folder uploads as volume => digitalocean spaces

**Comment by polRk** [comment]
O MY GOD !!
```
bash-5.0# ls
cache  empty  lib    local  lock   log    opt    run    spool  tmp    www    wwww
```

In my helm config file i have spelling error www**w**
FOUR **w**

I upload file in space. And its present
```
bash-5.0# ls wwww/html/wp-content/uploads/
backups.zip
```
It's my error (

**Comment by jbonnett92** [comment]
Perfect :)

**Comment by jbonnett92** [comment]
To be fair I have no idea how I missed that either!

**Comment by polRk** [comment]
I have another error

```
Unable to create directory uploads/2019/11. Is its parent directory writable by the server?
```

➜  kubernetes kubectl exec -it domain-ci-5c9999c57c-rnxkn bash
```
bash-5.0# ls -l wp-content
total 16
-rw-r--r--    1 root     root            30 Nov  4 15:33 index.php
drwxr-xr-x   12 root     root          4096 Nov  4 15:33 plugins
drwxr-xr-x    3 root     root          4096 Nov  4 15:33 themes
drwxr-xr-x    2 root     root          4096 Nov  4 15:34 uploads
```
How can i changed chmod ?

**Comment by jbonnett92** [comment]
Well you don't want to do that for security reasons, you need to log into the wordpress process owner or root and then create it.

**Comment by polRk** [comment]
> Well you don't want to do that for security reasons, you need to log into the wordpress process owner or root and then create it.

This comment will help me ?

mountOptions: "dir_mode=0777,file_mode=0777"
https://github.com/kubernetes/kubernetes/issues/54610#issuecomment-339662597

**Comment by jbonnett92** [comment]
You don't want to do that for security as I said.

**Comment by polRk** [comment]
> Well you don't want to do that for security reasons, you need to log into the wordpress process owner or root and then create it.

In my base wordpress image i change folder permissions.
https://github.com/polRk/docker-wordpress/blob/master/Dockerfile#L48

> You don't want to do that for security as I said.

I need it ) I want to uploads file. 
Or i can create initContainer
https://www.digitalocean.com/docs/kubernetes/how-to/add-volumes/#define-the-pod-1

**Comment by jbonnett92** [comment]
You don't need it, create files and folders under the wordpress owner or root.

**Comment by polRk** [comment]
> You don't need it, create files and folders under the wordpress owner or root.

It imposible. I need to upload a photo through the site

**Comment by jbonnett92** [comment]
How is it impossible?

**Comment by jbonnett92** [comment]
Check the owner and the owner group and use that.

**Comment by jbonnett92** [comment]
https://askubuntu.com/questions/175054/how-to-find-owner-and-group-of-a-directory

and then 

https://unix.stackexchange.com/questions/3568/how-to-switch-between-users-on-one-terminal

**Comment by jbonnett92** [comment]
So in this case the users is root and the group is root, so just use sudo before the command.

**Comment by jbonnett92** [comment]
https://www.cyberciti.biz/faq/linux-find-process-name/
https://unix.stackexchange.com/questions/284934/return-owner-of-process-given-pid

**Comment by polRk** [comment]
I use Docker image. It need  automatic change chmod or chown.

**Comment by jbonnett92** [comment]
A docker container is still Linux.

**Comment by polRk** [comment]
> A docker container is still Linux.

I change permisions https://github.com/polRk/docker-wordpress/blob/master/Dockerfile#L48
But as the result i got 
```
drwxr-xr-x    2 root     root          4096 Nov  4 15:56 uploads
```

**Comment by polRk** [comment]
> So in this case the users is root and the group is root, so just use sudo before the command.

Yes, you are right! I run php-fpm as root and i can read/write uploads folder

**Comment by AlirezaHaghshenas** [comment]
I get this error in the provisioner

```
I0516 12:37:14.712280       1 controller.go:1196] provision "default/private-configs" class "csi-s3": started
I0516 12:37:14.717203       1 event.go:209] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"private-configs", UID:"f7572a6d-63d2-478c-8362-9da370c86bf1", APIVersion:"v1", ResourceVersion:"1725", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "default/private-configs"
I0516 12:37:14.718166       1 controller.go:979] Final error received, removing PVC f7572a6d-63d2-478c-8362-9da370c86bf1 from claims in progress
W0516 12:37:14.718260       1 controller.go:886] Retrying syncing claim "f7572a6d-63d2-478c-8362-9da370c86bf1", failure 5
E0516 12:37:14.718368       1 controller.go:908] error syncing claim "f7572a6d-63d2-478c-8362-9da370c86bf1": failed to provision volume with StorageClass "csi-s3": rpc error: code = Unknown desc = failed to initialize S3 client: Endpoint:  does not follow ip address or domain name standards.
I0516 12:37:14.718462       1 event.go:209] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"private-configs", UID:"f7572a6d-63d2-478c-8362-9da370c86bf1", APIVersion:"v1", ResourceVersion:"1725", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "csi-s3": rpc error: code = Unknown desc = failed to initialize S3 client: Endpoint:  does not follow ip address or domain name standards.
```
in the secret, the endpoint is:

```
apiVersion: v1
kind: Secret
metadata:
  name: csi-s3-secret
  namespace: infra
stringData:
  accessKeyID: ReplaceThis
  secretAccessKey: ReplaceThis
  endpoint: https://fra1.digitaloceanspaces.com
  region: ""
```

I've also changed the namespaces to refer to `infra`, so I guess the problem is not here.

**Comment by favas0** [comment]
@AlirezaHaghshenas 
i am having the same error. Have you found a workaround ?

**Comment by alirezacads** [comment]
No, we moved on, worked around our problems using NFS and then took the time to refactor our code to do not require storage mounting.

**Comment by favas0** [comment]
Cool. Our deployment is slowed down when running it as stateless. Now looking into https://rook.io

**Comment by Berndinox** [comment]
To be honest, i would prefer using a plugin for storing photos in DO Spaces.
Using a plugin gives you the opportunity to rewrite the url, so visitors are loading the content directly from the do spaces.

**Comment by warrmr** [comment]
I am having the same problem as @AlirezaHaghshenas on digitalocean spaces. It doesnt seem to contact the DO spaces endpoint as I have even attempted to put in an invalid key/secert expecting it to throw an error about an invalid key but I just get the same error that it does not follow ip address or domain name standards. 

`Unknown desc = failed to initialize S3 client: Endpoint:  does not follow ip address or domain name standards.`

The endpoint i was using is http://nyc3.digitaloceanspaces.com I have tried ams3.digitaloceanspaces.com in all formats. http and https etc 

```
apiVersion: v1
kind: Secret
metadata:
    namespace: default
    name: csi-s3-secret
stringData:
    accessKeyID: Seceret
    secretAccessKey: Seceret
    endpoint: http://nyc3.digitaloceanspaces.com
    region: ""

```

**Comment by gre9ory** [comment]
Had this error as well but then it started to work fine - even having deployed everything in a custom namespace (so not `kube-system` or `default`).

Maybe these observations help someone:

- From looking at the code where the endpoint uri is parsed it is (at least right now) expected to provide the endpoint including scheme e.g. like this: `https://nyc3.digitaloceanspaces.com`

- It may also help/be required to add the `csi.storage` annotations to the StorageClass and take care the namespaces are correct:

```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: csi-s3
provisioner: ch.ctrox.csi.s3-driver
parameters:
  # specify which mounter to use
  # can be set to rclone, s3fs, goofys or s3backer
  mounter: goofys
  csi.storage.k8s.io/provisioner-secret-name: csi-s3-secret
  csi.storage.k8s.io/provisioner-secret-namespace: kube-system
  csi.storage.k8s.io/controller-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/controller-publish-secret-namespace: kube-system
  csi.storage.k8s.io/node-stage-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-stage-secret-namespace: kube-system
  csi.storage.k8s.io/node-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-publish-secret-namespace: kube-system`
```

**Comment by CallMeLaNN** [comment]
TLDR; Below is my working example and I can use rclone. Latest provided yaml in this repo doesn't work because permission for attacher need to [fix #70](https://github.com/ctrox/csi-s3/pull/70).

---

> I've been using DigitalOcean Space and use csi-s3(using rclone) to mount the S3 compatible storage and everything works fine

This comment from https://github.com/ctrox/csi-s3/pull/70 encourage me and I manage to make it work after try-an-error many times.

Initially from the latest README and example applied into my **latest DO managed Kubernetes v1.25 and Spaces**, I got this error:

```log
```

I manage to fix it after I apply the patch from https://github.com/ctrox/csi-s3/pull/70, but because I'm using `prefix` I got another error something like this in csi-s3 container:

```log
GRPC call: /csi.v1.Node/NodeStageVolume
GRPC error: The specified key does not exist.
```

It was reported here https://github.com/ctrox/csi-s3/issues/56 but no clue until I found the cause [my `prefix` contain path separator](...) So now it is working after I remove the path.

---

Here is the example:

```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: csi-s3-secret
  namespace: kube-system
stringData:
  accessKeyID: xxx
  secretAccessKey: xxx
  endpoint: https://sgp1.digitaloceanspaces.com
  region: ""
---
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: csi-s3
provisioner: ch.ctrox.csi.s3-driver
reclaimPolicy: Retain
parameters:
  mounter: rclone
  bucket: myexistingbucket
  usePrefix: "true"
  prefix: "mount-csi-s3"
  csi.storage.k8s.io/provisioner-secret-name: csi-s3-secret
  csi.storage.k8s.io/provisioner-secret-namespace: kube-system
  csi.storage.k8s.io/controller-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/controller-publish-secret-namespace: kube-system
  csi.storage.k8s.io/node-stage-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-stage-secret-namespace: kube-system
  csi.storage.k8s.io/node-publish-secret-name: csi-s3-secret
  csi.storage.k8s.io/node-publish-secret-namespace: kube-system
```
